
from dataclasses import dataclass
import typing
import typing_extensions

from flytekit.core.annotation import FlyteAnnotation

from latch.types.metadata import NextflowParameter
from latch.types.file import LatchFile
from latch.types.directory import LatchDir, LatchOutputDir

# Import these into your `__init__.py` file:
#
# from .parameters import generated_parameters

generated_parameters = {
    'study_name': NextflowParameter(
        type=str,
        default='study',
        section_title='Input/output options',
        description='A string to identify results in the output directory',
    ),
    'study_type': NextflowParameter(
        type=str,
        default='rnaseq',
        section_title=None,
        description='A string identifying the technology used to produce the data',
    ),
    'input': NextflowParameter(
        type=LatchFile,
        default=None,
        section_title=None,
        description='Path to comma-separated file containing information about the samples in the experiment.',
    ),
    'contrasts': NextflowParameter(
        type=LatchFile,
        default=None,
        section_title=None,
        description='A CSV file describing sample contrasts',
    ),
    'outdir': NextflowParameter(
        type=typing_extensions.Annotated[LatchDir, FlyteAnnotation({'output': True})],
        default=None,
        section_title=None,
        description='The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.',
    ),
    'study_abundance_type': NextflowParameter(
        type=str,
        default='counts',
        section_title=None,
        description='Type of abundance measure used, platform-dependent',
    ),
    'matrix': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title='Abundance values',
        description='TSV-format abundance matrix',
    ),
    'transcript_length_matrix': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='(RNA-seq only): optional transcript length matrix with samples and genes as the abundance matrix',
    ),
    'affy_cel_files_archive': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='Alternative to matrix: a compressed CEL files archive such as often found in GEO',
    ),
    'querygse': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='Use SOFT files from GEO by providing the GSE study identifier',
    ),
    'observations_id_col': NextflowParameter(
        type=str,
        default='sample',
        section_title='Observations (e.g. samples) options',
        description='Column in the samples sheet to be used as the primary sample identifier',
    ),
    'observations_type': NextflowParameter(
        type=str,
        default='sample',
        section_title=None,
        description='Type of observation',
    ),
    'observations_name_col': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Column in the sample sheet to be used as the display identifier for observations. If unset, will use value of --observations_id_col.',
    ),
    'features_id_col': NextflowParameter(
        type=str,
        default='gene_id',
        section_title='Features options',
        description='Feature ID attribute in the abundance table as well as in the GTF file (e.g. the gene_id field)',
    ),
    'features_name_col': NextflowParameter(
        type=str,
        default='gene_name',
        section_title=None,
        description='Feature name attribute in the abundance table as well as in the GTF file (e.g. the gene symbol field)',
    ),
    'features_type': NextflowParameter(
        type=str,
        default='gene',
        section_title=None,
        description="Type of feature we have, often 'gene'",
    ),
    'sizefactors_from_controls': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='When set, use the control features in scaling/ normalisation',
    ),
    'control_features': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='A text file listing technical features (e.g. spikes)',
    ),
    'features_metadata_cols': NextflowParameter(
        type=typing.Optional[str],
        default='gene_id,gene_name,gene_biotype',
        section_title=None,
        description='Comma-separated string, specifies feature metadata columns to be used for exploratory analysis, platform-specific',
    ),
    'features': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='This parameter allows you to supply your own feature annotations. These can often be automatically derived from the GTF used upstream for RNA-seq, or from the Bioconductor annotation package (for affy arrays). ',
    ),
    'features_gtf_feature_type': NextflowParameter(
        type=typing.Optional[str],
        default='transcript',
        section_title=None,
        description='Where a GTF file is supplied, which feature type to use',
    ),
    'features_gtf_table_first_field': NextflowParameter(
        type=typing.Optional[str],
        default='gene_id',
        section_title=None,
        description='Where a GTF file is supplied, which field should go first in the converted output table',
    ),
    'affy_file_name_col': NextflowParameter(
        type=typing.Optional[str],
        default='file',
        section_title='Affy input options',
        description='Column of the sample sheet containing the Affymetrix CEL file name',
    ),
    'affy_background': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='logical value. If TRUE, then background correct using RMA background correction.',
    ),
    'affy_bgversion': NextflowParameter(
        type=typing.Optional[int],
        default=2,
        section_title=None,
        description='integer value indicating which RMA background to use',
    ),
    'affy_destructive': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='logical value. If TRUE, then works on the PM matrix in place as much as possible, good for large datasets.',
    ),
    'affy_cdfname': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description="Used to specify the name of an alternative cdf package. If set to NULL, then the usual cdf package based on Affymetrix' mappings will be used.",
    ),
    'affy_build_annotation': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='logical value. If TRUE, a matrix of probe annotations will be derived.',
    ),
    'affy_rm_mask': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="should the spots marked as 'MASKS' set to NA?",
    ),
    'affy_rm_outliers': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="should the spots marked as 'OUTLIERS' set to NA?",
    ),
    'affy_rm_extra': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='if TRUE, then overrides what is in rm.mask and rm.oultiers.',
    ),
    'proteus_measurecol_prefix': NextflowParameter(
        type=typing.Optional[str],
        default='LFQ intensity',
        section_title='Proteus input options',
        description="Prefix of the column names of the MaxQuant proteingroups table in which the intensity values are saved; the prefix has to be followed by the sample names that are also found in the samplesheet. Default: 'LFQ intensity'; will search for both the prefix as entered and the prefix followed by one whitespace.",
    ),
    'proteus_norm_function': NextflowParameter(
        type=typing.Optional[str],
        default='normalizeMedian',
        section_title=None,
        description='Normalization function to use on the MaxQuant intensities.',
    ),
    'proteus_plotsd_method': NextflowParameter(
        type=typing.Optional[str],
        default='violin',
        section_title=None,
        description="Which method to use for plotting sample distributions of the MaxQuant intensities; one of 'violin', 'dist', 'box'.",
    ),
    'proteus_plotmv_loess': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Should a loess line be added to the plot of mean-variance relationship of the conditions? Default: true.',
    ),
    'proteus_palette_name': NextflowParameter(
        type=typing.Optional[str],
        default='Set1',
        section_title=None,
        description='Valid R palette name',
    ),
    'filtering_min_abundance': NextflowParameter(
        type=float,
        default=1.0,
        section_title='Filtering',
        description='Minimum abundance value',
    ),
    'filtering_min_samples': NextflowParameter(
        type=typing.Optional[float],
        default=1.0,
        section_title=None,
        description='Minimum observations that must pass the threshold to retain the row/ feature (e.g. gene).',
    ),
    'filtering_min_proportion': NextflowParameter(
        type=typing.Optional[float],
        default=None,
        section_title=None,
        description='A minimum proportion of observations, given as a number between 0 and 1, that must pass the threshold. Overrides  minimum_samples',
    ),
    'filtering_grouping_var': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='An optional grouping variable to be used to calculate a min_samples value',
    ),
    'filtering_min_proportion_not_na': NextflowParameter(
        type=typing.Optional[float],
        default=0.5,
        section_title=None,
        description='A minimum proportion of observations, given as a number between 0 and 1, that must have a value (not NA) to retain the row/ feature (e.g. gene).',
    ),
    'filtering_min_samples_not_na': NextflowParameter(
        type=typing.Optional[float],
        default=None,
        section_title=None,
        description='Minimum observations that must have a value (not NA) to retain the row/ feature (e.g. gene). Overrides filtering_min_proportion_not_na.',
    ),
    'exploratory_clustering_method': NextflowParameter(
        type=str,
        default='ward.D2',
        section_title='Exploratory analysis',
        description='Clustering method used in dendrogram creation',
    ),
    'exploratory_cor_method': NextflowParameter(
        type=str,
        default='spearman',
        section_title=None,
        description='Correlation method used in dendrogram creation',
    ),
    'exploratory_n_features': NextflowParameter(
        type=int,
        default=500,
        section_title=None,
        description='Number of features selected before certain exploratory analyses. If -1, will use all features.',
    ),
    'exploratory_whisker_distance': NextflowParameter(
        type=typing.Optional[float],
        default=1.5,
        section_title=None,
        description='Length of the whiskers in boxplots as multiple of IQR. Defaults to 1.5.',
    ),
    'exploratory_mad_threshold': NextflowParameter(
        type=typing.Optional[int],
        default=-5,
        section_title=None,
        description='Threshold on MAD score for outlier identification',
    ),
    'exploratory_main_variable': NextflowParameter(
        type=str,
        default='auto_pca',
        section_title=None,
        description="How should the main grouping variable be selected? 'auto_pca', 'contrasts', or a valid column name from the observations table.",
    ),
    'exploratory_log2_assays': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Of which assays to compute the log2 during exploratory analysis. Not necessary for maxquant data as this is controlled by the pipeline.',
    ),
    'exploratory_palette_name': NextflowParameter(
        type=str,
        default='Set1',
        section_title=None,
        description='Valid R palette name',
    ),
    'differential_file_suffix': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Differential analysis',
        description='Advanced option: the suffix associated tabular differential results tables. Will by default use the appropriate suffix according to the study_type.',
    ),
    'differential_feature_id_column': NextflowParameter(
        type=str,
        default='gene_id',
        section_title=None,
        description='The feature identifier column in differential results tables',
    ),
    'differential_fc_column': NextflowParameter(
        type=str,
        default='log2FoldChange',
        section_title=None,
        description='The fold change column in differential results tables',
    ),
    'differential_pval_column': NextflowParameter(
        type=typing.Optional[str],
        default='pvalue',
        section_title=None,
        description='The p value column in differential results tables',
    ),
    'differential_qval_column': NextflowParameter(
        type=str,
        default='padj',
        section_title=None,
        description='The q value column in differential results tables.',
    ),
    'differential_min_fold_change': NextflowParameter(
        type=float,
        default=2.0,
        section_title=None,
        description='Minimum fold change used to calculate differential feature numbers',
    ),
    'differential_max_pval': NextflowParameter(
        type=float,
        default=1.0,
        section_title=None,
        description='Maximum p value used to calculate differential feature numbers',
    ),
    'differential_max_qval': NextflowParameter(
        type=float,
        default=0.05,
        section_title=None,
        description='Maximum q value used to calculate differential feature numbers',
    ),
    'differential_feature_name_column': NextflowParameter(
        type=typing.Optional[str],
        default='gene_name',
        section_title=None,
        description='Where a features file (GTF) has been provided, what attributed to use to name features',
    ),
    'differential_foldchanges_logged': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Indicate whether or not fold changes are on the log scale (default is to assume they are)',
    ),
    'differential_palette_name': NextflowParameter(
        type=str,
        default='Set1',
        section_title=None,
        description='Valid R palette name',
    ),
    'differential_subset_to_contrast_samples': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='In differential analysis (DEseq2 or Limma), subset to the contrast samples before modelling variance?',
    ),
    'deseq2_test': NextflowParameter(
        type=typing.Optional[str],
        default='Wald',
        section_title='DESeq2 specific options (RNA-seq only)',
        description='`test` parameter passed to DESeq()',
    ),
    'deseq2_fit_type': NextflowParameter(
        type=typing.Optional[str],
        default='parametric',
        section_title=None,
        description='`fitType` parameter passed to DESeq()',
    ),
    'deseq2_sf_type': NextflowParameter(
        type=typing.Optional[str],
        default='ratio',
        section_title=None,
        description='`sfType` parameter passed to DESeq()',
    ),
    'deseq2_min_replicates_for_replace': NextflowParameter(
        type=typing.Optional[int],
        default=7,
        section_title=None,
        description="'minReplicatesForReplace' parameter passed to DESeq()",
    ),
    'deseq2_use_t': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='`useT` parameter passed to DESeq2',
    ),
    'deseq2_independent_filtering': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='`independentFiltering` parameter passed to results()',
    ),
    'deseq2_lfc_threshold': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='`lfcThreshold` parameter passed to results()',
    ),
    'deseq2_alt_hypothesis': NextflowParameter(
        type=typing.Optional[str],
        default='greaterAbs',
        section_title=None,
        description='`altHypothesis` parameter passed to results()',
    ),
    'deseq2_p_adjust_method': NextflowParameter(
        type=typing.Optional[str],
        default='BH',
        section_title=None,
        description='`pAdjustMethod` parameter passed to results()',
    ),
    'deseq2_alpha': NextflowParameter(
        type=typing.Optional[float],
        default=0.1,
        section_title=None,
        description='`alpha` parameter passed to results()',
    ),
    'deseq2_minmu': NextflowParameter(
        type=typing.Optional[float],
        default=0.5,
        section_title=None,
        description='`minmu` parameter passed to results()',
    ),
    'deseq2_vs_method': NextflowParameter(
        type=typing.Optional[str],
        default='vst',
        section_title=None,
        description='variance stabilisation method to use when making a variance stabilised matrix',
    ),
    'deseq2_shrink_lfc': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Shink fold changes in results?',
    ),
    'deseq2_cores': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description='Number of cores',
    ),
    'deseq2_vs_blind': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='`blind` parameter for rlog() and/ or vst()',
    ),
    'deseq2_vst_nsub': NextflowParameter(
        type=typing.Optional[int],
        default=1000,
        section_title=None,
        description='`nsub` parameter passed to vst()',
    ),
    'limma_ndups': NextflowParameter(
        type=typing.Optional[float],
        default=None,
        section_title='Limma specific options (microarray only)',
        description='passed to lmFit(), positive integer giving the number of times each distinct probe is printed on each array.',
    ),
    'limma_spacing': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='passed to lmFit(), positive integer giving the spacing between duplicate occurrences of the same probe, spacing=1 for consecutive rows.',
    ),
    'limma_block': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='Sample sheet column to be used to derive a vector or factor specifying a blocking variable on the arrays',
    ),
    'limma_correlation': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='passed to lmFit(), the inter-duplicate or inter-technical replicate correlation',
    ),
    'limma_method': NextflowParameter(
        type=typing.Optional[str],
        default='ls',
        section_title=None,
        description='passed to lmFit(), the fitting method',
    ),
    'limma_proportion': NextflowParameter(
        type=typing.Optional[float],
        default=0.01,
        section_title=None,
        description='passed to eBayes(), a numeric value between 0 and 1, assumed proportion of genes which are differentially expressed',
    ),
    'limma_trend': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='passed to eBayes(), logical, should an intensity-dependent trend be allowed for the prior variance?',
    ),
    'limma_robust': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='passed to eBayes(), logical, should the estimation of df.prior and var.prior be robustified against outlier sample variances?',
    ),
    'limma_stdev_coef_lim': NextflowParameter(
        type=typing.Optional[str],
        default='0.1,4',
        section_title=None,
        description='passed to eBayes, comma separated string of two values, assumed lower and upper limits for the standard deviation of log2-fold-changes for differentially expressed genes',
    ),
    'limma_winsor_tail_p': NextflowParameter(
        type=typing.Optional[str],
        default='0.05,0.1',
        section_title=None,
        description='passed to eBayes, comma separated string of length 1 or 2, giving left and right tail proportions of x to Winsorize. Used only when robust=TRUE.',
    ),
    'limma_lfc': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='passed to topTable(), minimum absolute log2-fold-change required',
    ),
    'limma_confint': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='passed to topTable(), logical, should confidence 95% intervals be output for logFC? Alternatively, can take a numeric value between zero and one specifying the confidence level required.',
    ),
    'limma_adjust_method': NextflowParameter(
        type=typing.Optional[str],
        default='BH',
        section_title=None,
        description='passed to topTable(), method used to adjust the p-values for multiple testing.',
    ),
    'limma_p_value': NextflowParameter(
        type=typing.Optional[float],
        default=1.0,
        section_title=None,
        description='cutoff value for adjusted p-values. Only genes with lower p-values are listed.',
    ),
    'gsea_run': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='GSEA',
        description='Set to run GSEA to infer differential gene sets in contrasts',
    ),
    'gsea_permute': NextflowParameter(
        type=typing.Optional[str],
        default='phenotype',
        section_title=None,
        description='Permutation type',
    ),
    'gsea_nperm': NextflowParameter(
        type=typing.Optional[int],
        default=1000,
        section_title=None,
        description='Number of permutations',
    ),
    'gsea_scoring_scheme': NextflowParameter(
        type=typing.Optional[str],
        default='weighted',
        section_title=None,
        description='Enrichment statistic',
    ),
    'gsea_metric': NextflowParameter(
        type=typing.Optional[str],
        default='Signal2Noise',
        section_title=None,
        description='Metric for ranking genes',
    ),
    'gsea_sort': NextflowParameter(
        type=typing.Optional[str],
        default='real',
        section_title=None,
        description='Gene list sorting mode',
    ),
    'gsea_order': NextflowParameter(
        type=typing.Optional[str],
        default='descending',
        section_title=None,
        description='Gene list ordering mode',
    ),
    'gsea_set_max': NextflowParameter(
        type=typing.Optional[int],
        default=500,
        section_title=None,
        description='Max size: exclude larger sets',
    ),
    'gsea_set_min': NextflowParameter(
        type=typing.Optional[int],
        default=15,
        section_title=None,
        description='Min size: exclude smaller sets',
    ),
    'gsea_norm': NextflowParameter(
        type=typing.Optional[str],
        default='meandiv',
        section_title=None,
        description='Normalisation mode',
    ),
    'gsea_rnd_type': NextflowParameter(
        type=typing.Optional[str],
        default='no_balance',
        section_title=None,
        description='Randomization mode',
    ),
    'gsea_make_sets': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Make detailed geneset report?',
    ),
    'gsea_median': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Use median for class metrics',
    ),
    'gsea_num': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title=None,
        description='Number of markers',
    ),
    'gsea_plot_top_x': NextflowParameter(
        type=typing.Optional[int],
        default=20,
        section_title=None,
        description='Plot graphs for the top sets of each phenotype',
    ),
    'gsea_rnd_seed': NextflowParameter(
        type=typing.Optional[str],
        default='timestamp',
        section_title=None,
        description='Seed for permutation',
    ),
    'gsea_save_rnd_lists': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Save random ranked lists',
    ),
    'gsea_zip_report': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Make a zipped file with all reports',
    ),
    'gprofiler2_run': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='gprofiler2',
        description='Set to run gprofiler2 and do a pathway enrichment analysis.',
    ),
    'gprofiler2_organism': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Short name of the organism that is analyzed, e.g. hsapiens for homo sapiens.',
    ),
    'gprofiler2_significant': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Should only significant enrichment results be considered?',
    ),
    'gprofiler2_measure_underrepresentation': NextflowParameter(
        type=typing.Optional[bool],
        default=False,
        section_title=None,
        description='Should underrepresentation be measured instead of overrepresentation?',
    ),
    'gprofiler2_correction_method': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='The method that should be used for multiple testing correction.',
    ),
    'gprofiler2_sources': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='On which source databases to run the gprofiler query',
    ),
    'gprofiler2_evcodes': NextflowParameter(
        type=typing.Optional[bool],
        default=False,
        section_title=None,
        description='Whether to include evcodes in the results.',
    ),
    'gprofiler2_max_qval': NextflowParameter(
        type=typing.Optional[float],
        default=0.05,
        section_title=None,
        description='Maximum q value used for significance testing.',
    ),
    'gprofiler2_token': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Token that should be used as a query.',
    ),
    'gprofiler2_background_file': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to CSV/TSV/TXT file that should be used as a background for the query; alternatively, 'auto' (default) or 'false'.",
    ),
    'gprofiler2_background_column': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Which column to use as gene IDs in the background matrix.',
    ),
    'gprofiler2_domain_scope': NextflowParameter(
        type=typing.Optional[str],
        default='annotated',
        section_title=None,
        description='How to calculate the statistical domain size.',
    ),
    'gprofiler2_min_diff': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description='How many genes must be differentially expressed in a pathway for it to be considered enriched? Default 1.',
    ),
    'gprofiler2_palette_name': NextflowParameter(
        type=typing.Optional[str],
        default='Blues',
        section_title=None,
        description='Valid R palette name',
    ),
    'shinyngs_build_app': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title='Shiny app settings',
        description='Should a Shiny app be built?',
    ),
    'shinyngs_deploy_to_shinyapps_io': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Should the app be deployed to shinyapps.io?',
    ),
    'shinyngs_shinyapps_account': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='Your shinyapps.io account name',
    ),
    'shinyngs_shinyapps_app_name': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='The name of the app to push to in your shinyapps.io account',
    ),
    'shinyngs_guess_unlog_matrices': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Should we guess the log status of matrices and unlog for the app?',
    ),
    'gene_sets_files': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title='Options related to gene set analysis',
        description='Gene sets in GMT or GMX-format; for GSEA: multiple comma-separated input files in either format are possible. For gprofiler2: A single file in GMT format is possible; this has lowest priority and will be overridden by --gprofiler2_token and --gprofiler2_organism.',
    ),
    'report_file': NextflowParameter(
        type=LatchFile,
        default=None,
        section_title='Reporting options',
        description='Rmd report template from which to create the pipeline report',
    ),
    'email': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Email address for completion summary.',
    ),
    'logo_file': NextflowParameter(
        type=str,
        default='${projectDir}/docs/images/nf-core-differentialabundance_logo_light.png',
        section_title=None,
        description='A logo to display in the report instead of the generic pipeline logo',
    ),
    'css_file': NextflowParameter(
        type=str,
        default='${projectDir}/assets/nf-core_style.css',
        section_title=None,
        description='CSS to use to style the output, in lieu of the default nf-core styling',
    ),
    'citations_file': NextflowParameter(
        type=typing.Optional[str],
        default='${projectDir}/CITATIONS.md',
        section_title=None,
        description='A markdown file containing citations to include in the fiinal report',
    ),
    'report_title': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='A title for reporting outputs',
    ),
    'report_author': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='An author for reporting outputs',
    ),
    'report_contributors': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Semicolon-separated string of contributor info that should be listed in the report.',
    ),
    'report_description': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='A description for reporting outputs',
    ),
    'report_scree': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Whether to generate a scree plot in the report',
    ),
    'report_round_digits': NextflowParameter(
        type=typing.Optional[int],
        default=4,
        section_title=None,
        description='To how many digits should numeric output in different modules be rounded? If -1, will not round.',
    ),
    'genome': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Reference genome options',
        description='Name of iGenomes reference.',
    ),
    'gtf': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Genome annotation file in GTF format',
    ),
    'publish_dir_mode': NextflowParameter(
        type=typing.Optional[str],
        default='copy',
        section_title='Generic options',
        description='Method used to save pipeline results to output directory.',
    ),
    'email_on_fail': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Email address for completion summary, only when pipeline fails.',
    ),
    'validate_params': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Boolean whether to validate parameters against the schema at runtime',
    ),
}

